sw/device/tests/pwm_smoketest.c

```diff
@@ -0,0 +1,104 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_gpio.h"
+#include "sw/device/lib/dif/dif_pinmux.h"
+#include "sw/device/lib/dif/dif_pwm.h"
+#include "sw/device/lib/runtime/hart.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/lib/testing/test_framework/ottf_utils.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+static volatile const uint8_t kClocksHz[] = {20, 50, 100, 200};
+static volatile const uint8_t kDutyCycles[] = {10, 30, 50, 70, 90};
```

Comment:
If the duty cycles were chosen to by power-of-two values rather than decimal values (e.g. fixed-point fractions of 0x10000 as per the IP block specification) then the errors would be smaller.
I realize that the interface provided by the DIF mismatches that of the register-level API and hardware implementation, but the precision would be improved by sticking duty cycles that are integral numbers in base 2 (or 16).